jib-core/src/main/java/com/google/cloud/tools/jib/builder/BuildTarballAndLoadDockerStep.java

```diff
@@ -0,0 +1,125 @@
+import com.google.cloud.tools.jib.image.json.ImageToJsonTranslator;
+import com.google.cloud.tools.jib.tar.TarStreamBuilder;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import java.io.BufferedOutputStream;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import org.apache.commons.compress.archivers.tar.TarArchiveEntry;
+
+class BuildTarballAndLoadDockerStep implements Callable<Void> {
+
+  private final BuildConfiguration buildConfiguration;
+  private final ListeningExecutorService listeningExecutorService;
+  private final ListenableFuture<List<ListenableFuture<CachedLayer>>>
+      pullBaseImageLayerFuturesFuture;
+  private final List<ListenableFuture<CachedLayer>> buildApplicationLayerFutures;
+  private final ListenableFuture<ListenableFuture<Blob>> buildConfigurationFutureFuture;
+
+  BuildTarballAndLoadDockerStep(
+      BuildConfiguration buildConfiguration,
+      ListeningExecutorService listeningExecutorService,
+      ListenableFuture<List<ListenableFuture<CachedLayer>>> pullBaseImageLayerFuturesFuture,
+      List<ListenableFuture<CachedLayer>> buildApplicationLayerFutures,
+      ListenableFuture<ListenableFuture<Blob>> buildConfigurationFutureFuture) {
+    this.buildConfiguration = buildConfiguration;
+    this.listeningExecutorService = listeningExecutorService;
+    this.pullBaseImageLayerFuturesFuture = pullBaseImageLayerFuturesFuture;
+    this.buildApplicationLayerFutures = buildApplicationLayerFutures;
+    this.buildConfigurationFutureFuture = buildConfigurationFutureFuture;
+  }
+
+  /**
+   * Depends on {@code pullBaseImageLayerFuturesFuture} and {@code buildConfigurationFutureFuture}.
+   */
+  @Override
+  public Void call() throws ExecutionException, InterruptedException {
+    List<ListenableFuture<?>> dependencies = new ArrayList<>();
+    dependencies.addAll(NonBlockingFutures.get(pullBaseImageLayerFuturesFuture));
+    dependencies.addAll(buildApplicationLayerFutures);
+    dependencies.add(NonBlockingFutures.get(buildConfigurationFutureFuture));
+    return Futures.whenAllComplete(dependencies)
+        .call(this::afterPushBaseImageLayerFuturesFuture, listeningExecutorService)
+        .get();
+  }
+
+  /**
+   * Depends on {@code pushAuthorizationFuture}, {@code pushBaseImageLayerFuturesFuture.get()},
+   * {@code pushApplicationLayerFutures}, and (@code
+   * containerConfigurationBlobDescriptorFutureFuture.get()}.
+   */
+  private Void afterPushBaseImageLayerFuturesFuture()
+      throws ExecutionException, InterruptedException, IOException {
+    // Add layers to image tarball
+    TarStreamBuilder builder = new TarStreamBuilder();
```

Comment:
Prob good to name this fully `tarStreamBuilder` since by like line 120, it might be hard to remember this was a `TarStreamBuilder`.